sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala

```diff
@@ -1196,49 +1174,101 @@ class SQLContext(@transient val sparkContext: SparkContext)
   // Register a succesfully instantiatd context to the singleton. This should be at the end of
   // the class definition so that the singleton is updated only if there is no exception in the
   // construction of the instance.
-  SQLContext.setLastInstantiatedContext(self)
+  sparkContext.addSparkListener(new SparkListener{
+    override def onApplicationEnd(applicationEnd: SparkListenerApplicationEnd) {
+      SQLContext.clearTheInstantiatedContext(self)
+    }
+  })
+
+  SQLContext.setInstantiatedContext(self)
 }
 
 /**
  * This SQLContext object contains utility functions to create a singleton SQLContext instance,
- * or to get the last created SQLContext instance.
+ * or to get the created SQLContext instance.
  */
 object SQLContext {
 
   private val INSTANTIATION_LOCK = new Object()
 
   /**
-   * Reference to the last created SQLContext.
+   * The active SQLContext for the current thread.
    */
-  @transient private val lastInstantiatedContext = new AtomicReference[SQLContext]()
+  private val activeContext: InheritableThreadLocal[SQLContext] =
+    new InheritableThreadLocal[SQLContext]
+
+  /**
+   * Reference to the created SQLContext.
+   */
+  @transient private val instantiatedContext = new AtomicReference[SQLContext]()
```

Comment:
is it possible to have just one of these? It seems currently that we only use `activeContext` if the user explicitly called `setActive`. Can we just use the same `AtomicReference` for `setActive`?